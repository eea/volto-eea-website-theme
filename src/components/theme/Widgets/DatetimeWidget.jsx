import React from 'react';
import cx from 'classnames';
import { useSelector } from 'react-redux';
import { toBackendLang } from '@plone/volto/helpers';
import { formatDate } from '@plone/volto/helpers/Utils/Date';
import config from '@plone/volto/registry';

export const DatetimeWidget = ({ value, children, className }) => {
  const lang = useSelector((state) => state.intl.locale);
  const backendLang = toBackendLang(lang);
  const locale =
    backendLang === 'en' ? config.settings.dateLocale : backendLang;
  const formatOptions = {
    date: value,
    format: {
      year: 'numeric',
      month: 'short',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
    },
    locale,
    includeTime: true,
    formatToParts: true,
  };

  let formattedParts = formatDate(formatOptions);

  const formattedDate = formattedParts
    .map((part) => {
      if (part.type === 'literal' && part.value === ', ') {
        return ' ';
      }
      return part.value;
    })
    .join('');

  return value ? (
    <span className={cx(className, 'datetime', 'widget')}>
      {children ? children(formattedDate) : formattedDate}
    </span>
  ) : (
    ''
  );
};
