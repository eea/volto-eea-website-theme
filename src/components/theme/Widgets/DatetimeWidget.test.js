import React from 'react';
import renderer from 'react-test-renderer';
import { DatetimeWidget } from './DatetimeWidget';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureStore([thunk]);

const store = mockStore({
  intl: {
    locale: 'en-gb',
    messages: {},
  },
});

describe('DatetimeWidget', () => {
  it('renders an empty datetime view widget component', () => {
    const component = renderer.create(
      <Provider store={store}>
        <DatetimeWidget />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a datetime view widget component with a date and time', () => {
    const component = renderer.create(
      <Provider store={store}>
        <DatetimeWidget className="metadata" value="2024-09-05T15:34:00" />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a datetime view widget component with children formatting', () => {
    const component = renderer.create(
      <Provider store={store}>
        <DatetimeWidget className="metadata" value="2024-09-05T15:34:00">
          {(formattedDate) => <strong>{formattedDate}</strong>}
        </DatetimeWidget>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('removes the comma in the formatted date and shows correct time', () => {
    const component = renderer.create(
      <Provider store={store}>
        <DatetimeWidget className="metadata" value="2024-09-05T15:34:00" />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();

    const instance = component.root;
    const span = instance.findByType('span');
    expect(span.props.children).toContain('05 Sept 2024 15:34');
  });
});
