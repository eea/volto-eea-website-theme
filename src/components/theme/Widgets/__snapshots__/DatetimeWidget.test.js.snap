// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatetimeWidget renders a date view widget component 1`] = `
<span
  className="metadata datetime widget"
>
  04 Aug 2020
</span>
`;

exports[`DatetimeWidget renders a date view widget component with children 1`] = `
<span
  className="metadata datetime widget"
>
  <strong>
    04 Aug 2020
  </strong>
</span>
`;

exports[`DatetimeWidget renders a date view widget component with custom format 1`] = `
<span
  className="metadata datetime widget"
>
  04 Aug 2020
</span>
`;

exports[`DatetimeWidget renders an empty date view widget component 1`] = `""`;
